languages/python/oso/oso/oso.py

```diff
@@ -80,12 +157,75 @@ def get_allowed_actions(self, actor, resource, allow_wildcard=False) -> list:
 
+    def authorize_field(self, actor, action, resource, field):
+        """Ensure that ``actor`` is allowed to perform ``action`` on a given
+        ``resource``'s ``field``.
+
+        If the action is permitted by an ``allow_field`` rule in the policy,
+        then this method returns ``None``. If the action is not permitted by the
+        policy, this method will raise a ``ForbiddenError``.
+
+        :param actor: The actor performing the request.
+        :param action: The action the actor is attempting to perform on the
+            field.
+        :param resource: The resource being accessed.
+        :param field: The name of the field being accessed.
+        """
+        if not self.query_rule_once("allow_field", actor, action, resource, field):
+            raise self._get_error(False)
+
+    def authorized_fields(
+        self, actor, action, resource, allow_wildcard=False
+    ) -> List[Any]:
+        """Determine the fields of ``resource`` on which ``actor`` is allowed to
+        perform  ``action``.
+
+        Uses ``allow_field`` rules in the policy to find all allowed fields.
+
+        :param actor: The actor for whom to collect allowed fields.
+        :param action: The action being taken on the field.
+        :param resource: The resource being accessed.
+
+        :param allow_wildcard: Flag to determine behavior if the policy \
+            includes a wildcard field. E.g., a rule allowing any field: \
+            ``allow_field(_actor, _action, _resource, _field)``. If ``True``, the \
+            method will return ``["*"]``, if ``False``, the method will raise an \
+            exception.
+
+        :type allow_wildcard: bool
+
+        :return: A list of the unique allowed fields.
+        """
+        results = self.query_rule(
+            "allow_field", actor, action, resource, Variable("field")
+        )
+        fields = set()
+        for result in results:
+            field = result.get("bindings").get("field")
+            if isinstance(field, Variable):
+                if not allow_wildcard:
+                    raise exceptions.OsoError(
+                        """The result of authorized_fields() contained an
+                        "unconstrained" field that could represent any
+                        field, but allow_wildcard was set to False. To fix,
+                        set allow_wildcard to True and compare with the "*"
+                        string."""
+                    )
+                else:
+                    return ["*"]
+            fields.add(field)
+
+        return list(fields)
+
     def _print_polar_log_message(self):
         if os.environ.get("POLAR_LOG", None):
             print(
                 "Polar tracing enabled. Get help with "
                 + "traces from our engineering team: https://help.osohq.com/trace"
             )
 
+    def _default_get_error(self, is_not_found):
+        return NotFoundError() if is_not_found else ForbiddenError()
```

Comment:
Same feedback as https://github.com/osohq/oso/pull/1102#discussion_r696307012 re: using a boolean to flip between two error variants

Reply:
Yep, making that change now to be consistent 👍